sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -401,6 +381,53 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   }
 
   /**
+   * :: Experimental ::
+   * Starts the execution of the streaming query, which will continually send results to the given
+   * [[ForeachWriter]] as new data arrives. The returned [[ContinuousQuery]] object can be used to
+   * interact with the stream.
+   *
+   * @since 2.0.0
+   */
+  @Experimental
+  def foreach(writer: ForeachWriter[T]): ContinuousQuery = {
+    assertNotBucketed("foreach")
+    assertStreaming(
+      "foreach() on streaming Datasets and DataFrames can only be called on continuous queries")
+
+    val queryName = extraOptions.getOrElse("queryName", StreamExecution.nextName)
+    val sink = new ForeachSink[T](ds.sparkSession.sparkContext.clean(writer))(ds.exprEnc)
+    df.sparkSession.sessionState.continuousQueryManager.startQuery(
+      queryName,
+      getCheckpointLocation(queryName, required = false),
+      df,
+      sink,
+      outputMode,
+      trigger)
+  }
+
+  /**
+   * Returns the checkpointLocation for a query. If `required` is `true` but the checkpoint
+   * location is not set, [[AnalysisException]] will be thrown. If `required` is `false`, a temp
+   * folder will be created if the checkpoint location is not set.
+   */
+  private def getCheckpointLocation(queryName: String, required: Boolean): String = {
```

Comment:
The semantics of this method is very confusing. `required` implies that it will throw error if there is not checkpoint location set. Its not intuitive that when it is not required it creates a temp checkpoint dir. Furthermore it creates temp one named on memory. Does not make sense. 

Cleaner for this to return an Option[String] and `required` --> `failIfNotSet'.